lib/stages/Transpose.cpp

```diff
@@ -0,0 +1,231 @@
+                     bufferContainer& buffer_container) :
+    Stage(config, unique_name, buffer_container, std::bind(&Transpose::main_thread, this)),
+    in_buf(get_buffer("in_buf")),
+    frame_id(in_buf) {
+
+    register_consumer(in_buf, unique_name.c_str());
+
+    // Get file path to write to
+    filename = config.get<std::string>(unique_name, "outfile");
+
+    // Get a timeout for communication with broker
+    timeout =
+        std::chrono::duration<float>(config.get_default<float>(unique_name, "comet_timeout", 60.));
+
+    // Collect some metadata. The rest is requested from the datasetManager,
+    // once we received the first frame.
+    metadata["archive_version"] = "3.1.0";
+    metadata["notes"] = "";
+    metadata["git_version_tag"] = get_git_commit_hash();
+    metadata["system_user"] = get_username();
+    metadata["collection_server"] = get_hostname();
+}
+
+void Transpose::main_thread() {
+
+    // frameID frame_id(in_buf);
+    uint32_t frames_so_far = 0;
+    // frequency and time indices within chunk
+    uint32_t fi = 0;
+    uint32_t ti = 0;
+    // offset for copying into buffer
+    uint32_t offset = 0;
+
+    // The dataset ID we read from the frame
+    dset_id_t ds_id;
+    // The dataset ID of the state without the time axis
+    dset_id_t base_ds_id;
+    // String formatted dataset ID to be written to the file
+    std::string ds_id_str;
+
+    size_t frame_size = 0;
+
+    // wait for a non-empty frame to get dataset ID from
+    uint32_t num_empty_skip = 0;
+    while (true) {
+        // Wait for a frame in the input buffer in order to get the dataset ID
+        if ((wait_for_full_frame(in_buf, unique_name.c_str(), frame_id)) == nullptr) {
+            return;
+        }
+
+        // Get the frame size to publish metrics later
+        uint64_t fpga_seq_total;
+        dset_id_t frame_ds_id;
+        std::tie(frame_size, fpga_seq_total, frame_ds_id) = get_frame_data();
+
+        // If the frame is empty, release the buffer and continue
+        if (fpga_seq_total == 0 && frame_ds_id == dset_id_t::null) {
+            DEBUG("Got empty frame ({:d}).", frame_id);
+            mark_frame_empty(in_buf, unique_name.c_str(), frame_id++);
+            num_empty_skip++;
+        } else {
+            ds_id = frame_ds_id;
+            break;
+        }
+    }
+
+    if (num_empty_skip > 0) {
+        INFO("Found {:d} empty frames at the start of the file.", num_empty_skip);
+    }
+
+    if (!get_dataset_state(ds_id)) {
+        FATAL_ERROR("Couldn't find ancestor of dataset {}. "
+                    "Make sure there is a stage upstream in the config, that adds the dataset "
+                    "states.\nExiting...",
+                    ds_id);
+        return;
+    }
+
+    // Get the original dataset ID (before adding time axis)
+    base_ds_id = base_dset(ds_id);
+
+    found_flags = std::vector<bool>(write_t, false);
+
+    // Create HDF5 file
+    create_hdf5_file();
```

Comment:
Can this fail? It's an IO operation so it presumably can, but I don't really see any way that's being dealt with.

Reply:
So this just creates a `HFBFileArchive`:
```
void HFBTranspose::create_hdf5_file() {
    // Create HDF5 file
    file = std::unique_ptr<HFBFileArchive>(
        new HFBFileArchive(filename, metadata, times, freqs, beams, sub_freqs, chunk,
                           kotekan::logLevel(_member_log_level)));                                                                                                                                                 
}
```
There were no checks in `VisTranspose` for it. Should I check if it is `NULL`?